bauble/src/value/mod.rs

```diff
@@ -0,0 +1,723 @@
+        Self {
+            ctx,
+            uses: HashMap::default(),
+        }
+    }
+    fn add_ref(
+        &mut self,
+        ident: String,
+        reference: Reference,
+    ) -> std::result::Result<(), ConversionError> {
+        let r = self.uses.entry(ident).or_default();
+
+        *r = r.clone().add(reference)?;
+
+        Ok(())
+    }
+
+    fn add<U: AssetContext>(&mut self, symbols: Symbols<U>) {
+        self.uses.extend(symbols.uses)
+    }
+
+    fn add_use(&mut self, use_: &Use) -> Result<()> {
+        fn add_use_inner<C: AssetContext>(
+            this: &mut Symbols<C>,
+            leading: String,
+            end: &Spanned<PathTreeEnd>,
+        ) -> Result<()> {
+            match &end.value {
+                PathTreeEnd::Group(g) => {
+                    for node in g {
+                        let mut leading = leading.clone();
+                        for s in &node.leading.value {
+                            leading.push_str(&s.value);
+                            leading.push_str("::");
+                        }
+                        add_use_inner(this, leading, &node.end)?;
+                    }
+                }
+                PathTreeEnd::Everything => {
+                    if let Some(uses) = this.ctx.all_in(&leading) {
+                        for (ident, reference) in uses {
+                            this.add_ref(ident, reference)
+                                .map_err(|e| e.span(end.span))?;
+                        }
+                    } else {
+                        return Err(ConversionError::ModuleNotFound.span(end.span));
+                    }
+                }
+                PathTreeEnd::PathEnd(PathEnd::Ident(ident)) => {
+                    let path = format!("{leading}{ident}");
+                    if let Some(reference) = this.ctx.get_ref(&path) {
+                        this.add_ref(ident.value.clone(), reference)
+                            .map_err(|e| e.span(ident.span))?;
+                    } else {
+                        return Err(ConversionError::ModuleNotFound.span(end.span));
+                    }
+                }
+                PathTreeEnd::PathEnd(PathEnd::WithIdent(ident)) => {
+                    if let Some(reference) = this.ctx.with_ident(&leading.trim_end_matches(':'), ident.as_str()) {
+                        this.add_ref(ident.value.clone(), reference)
+                            .map_err(|e| e.span(ident.span))?;
+                    } else {
+                        return Err(ConversionError::ModuleNotFound.span(end.span));
+                    }
+                }
+            }
+            Ok(())
+        }
+
+        let mut leading = String::new();
+        for l in use_.leading.iter() {
+            leading.push_str(l);
+            leading.push_str("::");
+        }
+        add_use_inner(self, leading, &use_.end)
+    }
+
+    fn try_resolve_macro<'a>(&'a self, ident: &str) -> Option<(&'a str, Option<Val>)> {
```

Comment:
Why is this called `try_resolve_macro()`?

Reply:
I think at the point I wrote it we hadn't come up with a name for it. And I had used macro when discussing about those items which are now prefixed with `copy`. But I renamed it to `try_resolve_copy`